src/sys/unix/uds/mod.rs

```diff
@@ -1,164 +1,167 @@
+
+pub(crate) mod datagram;
+pub(crate) mod listener;
+pub(crate) mod stream;
+
+const UNNAMED_ADDRESS: &[u8] = &[];
 
 /// Get the `sun_path` field offset of `sockaddr_un` for the target OS.
 ///
 /// On Linux, this function equates to the same value as
 /// `size_of::<sa_family_t>()`, but some other implementations include
 /// other fields before `sun_path`, so the expression more portably
 /// describes the size of the address structure.
-pub(in crate::sys) fn path_offset(sockaddr: &libc::sockaddr_un) -> usize {
+fn path_offset(sockaddr: &libc::sockaddr_un) -> usize {
     let base = sockaddr as *const _ as usize;
     let path = &sockaddr.sun_path as *const _ as usize;
     path - base
 }
 
-cfg_os_poll! {
-    use std::cmp::Ordering;
-    use std::os::unix::io::{RawFd, FromRawFd};
-    use std::{io, mem};
-
-    pub(crate) mod datagram;
-    pub(crate) mod listener;
-    pub(crate) mod stream;
-
-    pub(in crate::sys) fn socket_addr(bytes: &[u8]) -> io::Result<(libc::sockaddr_un, libc::socklen_t)> {
-        let sockaddr = mem::MaybeUninit::<libc::sockaddr_un>::zeroed();
-
-        // This is safe to assume because a `libc::sockaddr_un` filled with `0`
-        // bytes is properly initialized.
-        //
-        // `0` is a valid value for `sockaddr_un::sun_family`; it is
-        // `libc::AF_UNSPEC`.
-        //
-        // `[0; 108]` is a valid value for `sockaddr_un::sun_path`; it begins an
-        // abstract path.
-        let mut sockaddr = unsafe { sockaddr.assume_init() };
-
-        sockaddr.sun_family = libc::AF_UNIX as libc::sa_family_t;
-
-        match (bytes.first(), bytes.len().cmp(&sockaddr.sun_path.len())) {
-            // Abstract paths don't need a null terminator
-            (Some(&0), Ordering::Greater) => {
-                return Err(io::Error::new(
-                    io::ErrorKind::InvalidInput,
-                    "path must be no longer than libc::sockaddr_un.sun_path",
-                ));
+/// Converts a Rust `SocketAddr` into the system representation.
+fn unix_addr(address: &SocketAddr) -> (libc::sockaddr_un, libc::socklen_t) {
+    let sockaddr = mem::MaybeUninit::<libc::sockaddr_un>::zeroed();
+
+    // This is safe to assume because a `libc::sockaddr_un` filled with `0`
+    // bytes is properly initialized.
+    //
+    // `0` is a valid value for `sockaddr_un::sun_family`; it is
+    // `libc::AF_UNSPEC`.
+    //
+    // `[0; 108]` is a valid value for `sockaddr_un::sun_path`; it begins an
+    // abstract path.
+    let mut sockaddr = unsafe { sockaddr.assume_init() };
+
+    sockaddr.sun_family = libc::AF_UNIX as libc::sa_family_t;
+
+    #[allow(unused_mut)] // Only used with abstract namespaces.
+    let mut offset = 0;
+    let addr = match address.as_pathname() {
+        Some(path) => path.as_os_str().as_bytes(),
+        #[cfg(target_os = "linux")]
+        None => match address.as_abstract_name() {
+            Some(name) => {
+                offset += 1;
+                name
```

Comment:
Note that technically this change does not mean that the MSRV cannot be lower than 1.70.0. You can use a build script to detect the rustc version and only include this branch if the rustc is at least 1.70.0. This works because the user cannot construct a `SocketAddr` with an abstract name on older rustc.

We used this technique in Tokio to support the IO safety traits before our MSRV supported them.

Reply:
I know it's increasing the MSRV, that was the goal for Mio v1.

I'm not going to mess around with a build script. I've seen it in the log and libc crates as well as a few others and it's always been a pain to maintain them. Mio v1 is simply going to have a higher MSRV.